File: Project/java/src/main/java/com/basanta/todo/TaskManager.java
package com.basanta.todo;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Task operations: CRUD and filtering (language-specific: Optional, streams, Mutex).
 */
public final class TaskManager {
    private static final Mutex mutex = new Mutex();

    private static String generateId() {
        return "task_" + System.currentTimeMillis() + "_" + Long.toHexString(Double.doubleToLongBits(Math.random()));
    }

    public static Result<Task> create(String title, String description, String category, String status,
                                      int priority, String assignee, String createdBy) {
        String cat = category != null ? category : Category.OTHER;
        String st = status != null ? status : TaskStatus.PENDING;
        User.ValidationResult vr = Task.validateTaskInput(title, description, cat, st, priority);
        if (!vr.valid) return Result.fail(vr.error);

        return mutex.runExclusive(() -> {
            try {
                List<Task> tasks = new ArrayList<>(TaskRepository.loadTasks());
                Task task = Task.create(generateId(), title != null ? title.trim() : "",
                    description != null ? description.trim() : "", cat, st, priority,
                    assignee != null ? assignee.trim() : "", createdBy);
                tasks.add(task);
                TaskRepository.saveTasks(tasks);
                return Result.ok(task);
            } catch (IOException e) {
                return Result.fail(e.getMessage());
            }
        });
    }

    public static Optional<Task> getById(String taskId) {
        return mutex.runExclusive(() -> {
            try {
                return TaskRepository.loadTasks().stream()
                    .filter(t -> t.getId().equals(taskId))
                    .findFirst();
            } catch (IOException e) {
                return Optional.empty();
            }
        });
    }

    /** List with optional filters (streams for filtering). */
    public static List<Task> list(String category, String status, String assignee, String createdBy) {
        return mutex.runExclusive(() -> {
            try {
                List<Task> tasks = new ArrayList<>(TaskRepository.loadTasks());
                if (category != null && !category.isBlank())
                    tasks = tasks.stream().filter(t -> category.equals(t.getCategory())).toList();
                if (status != null && !status.isBlank())
                    tasks = tasks.stream().filter(t -> status.equals(t.getStatus())).toList();
                if (assignee != null && !assignee.isBlank())
                    tasks = tasks.stream().filter(t -> assignee.equals(t.getAssignee())).toList();
                if (createdBy != null && !createdBy.isBlank())
                    tasks = tasks.stream().filter(t -> createdBy.equals(t.getCreatedBy())).toList();
                return new ArrayList<>(tasks);
            } catch (IOException e) {
                return List.of();
            }
        });
    }

    public static Result<Task> update(String taskId, String title, String description, String category, String status, Integer priority, String assignee) {
        if (category != null && !Category.isValidCategory(category)) return Result.fail("Invalid category");
        if (status != null && !TaskStatus.isValidStatus(status)) return Result.fail("Invalid status");

        return mutex.runExclusive(() -> {
            try {
                List<Task> tasks = new ArrayList<>(TaskRepository.loadTasks());
                int idx = -1;
                for (int i = 0; i < tasks.size(); i++) {
                    if (tasks.get(i).getId().equals(taskId)) { idx = i; break; }
                }
                if (idx == -1) return Result.fail("Task not found");
                Task t = tasks.get(idx);
                tasks.set(idx, t.withUpdates(title, description, category, status, priority, assignee));
                TaskRepository.saveTasks(tasks);
                return Result.ok(tasks.get(idx));
            } catch (IOException e) {
                return Result.fail(e.getMessage());
            }
        });
    }

    public static Result<Void> remove(String taskId) {
        return mutex.runExclusive(() -> {
            try {
                List<Task> tasks = new ArrayList<>(TaskRepository.loadTasks());
                boolean removed = tasks.removeIf(t -> t.getId().equals(taskId));
                if (!removed) return Result.fail("Task not found");
                TaskRepository.saveTasks(tasks);
                return Result.ok(null);
            } catch (IOException e) {
                return Result.fail(e.getMessage());
            }
        });
    }
}

File: Project/javaScript/src/TaskManager.js
/**
 * Task operations: async CRUD, filtering by category/status/assignee.
 * Uses Mutex for critical sections and TaskRepository for persistence.
 */

import { Mutex } from './Mutex.js';
import * as TaskRepo from './TaskRepository.js';
import { createTask, updateTask, validateTaskInput } from './Task.js';
import { PENDING } from './TaskStatus.js';

const mutex = new Mutex();

/**
 * Generate a simple unique id.
 * @returns {string}
 */
function generateId() {
  return `task_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
}

/**
 * Create a task (locked).
 * @param {Object} opts
 * @param {string} opts.title
 * @param {string} [opts.description]
 * @param {string} opts.category
 * @param {string} [opts.status]
 * @param {number} [opts.priority]
 * @param {string} [opts.assignee]
 * @param {string} opts.createdBy
 * @returns {Promise<{ success: boolean, task?: import('./Task.js').Task, error?: string }>}
 */
export async function create(opts) {
  const validation = validateTaskInput({
    title: opts.title,
    description: opts.description,
    category: opts.category,
    status: opts.status ?? PENDING,
    priority: opts.priority,
  });
  if (!validation.valid) {
    return { success: false, error: validation.error };
  }

  return mutex.runExclusive(async () => {
    const tasks = await TaskRepo.loadTasks();
    const task = createTask({
      id: generateId(),
      title: opts.title.trim(),
      description: opts.description?.trim() ?? '',
      category: opts.category,
      status: opts.status ?? PENDING,
      priority: opts.priority ?? 0,
      assignee: opts.assignee?.trim() ?? '',
      createdBy: opts.createdBy,
    });
    tasks.push(task);
    await TaskRepo.saveTasks(tasks);
    return { success: true, task };
  });
}

/**
 * Get task by id.
 * @param {string} taskId
 * @returns {Promise<import('./Task.js').Task | null>}
 */
export async function getById(taskId) {
  return mutex.runExclusive(async () => {
    const tasks = await TaskRepo.loadTasks();
    return tasks.find((t) => t.id === taskId) ?? null;
  });
}

/**
 * List tasks with optional filters.
 * @param {Object} [filters]
 * @param {string} [filters.category]
 * @param {string} [filters.status]
 * @param {string} [filters.assignee]
 * @param {string} [filters.createdBy]
 * @returns {Promise<import('./Task.js').Task[]>}
 */
export async function list(filters = {}) {
  return mutex.runExclusive(async () => {
    let tasks = await TaskRepo.loadTasks();
    if (filters.category) tasks = tasks.filter((t) => t.category === filters.category);
    if (filters.status) tasks = tasks.filter((t) => t.status === filters.status);
    if (filters.assignee) tasks = tasks.filter((t) => t.assignee === filters.assignee);
    if (filters.createdBy) tasks = tasks.filter((t) => t.createdBy === filters.createdBy);
    return tasks;
  });
}

/**
 * Update a task (locked).
 * @param {string} taskId
 * @param {Partial<Pick<import('./Task.js').Task, 'title'|'description'|'category'|'status'|'priority'|'assignee'>>} updates
 * @returns {Promise<{ success: boolean, task?: import('./Task.js').Task, error?: string }>}
 */
export async function update(taskId, updates) {
  if (updates.category !== undefined && !(await import('./Category.js')).isValidCategory(updates.category)) {
    return { success: false, error: 'Invalid category' };
  }
  if (updates.status !== undefined && !(await import('./TaskStatus.js')).isValidStatus(updates.status)) {
    return { success: false, error: 'Invalid status' };
  }

  return mutex.runExclusive(async () => {
    const tasks = await TaskRepo.loadTasks();
    const idx = tasks.findIndex((t) => t.id === taskId);
    if (idx === -1) return { success: false, error: 'Task not found' };
    tasks[idx] = updateTask(tasks[idx], updates);
    await TaskRepo.saveTasks(tasks);
    return { success: true, task: tasks[idx] };
  });
}

/**
 * Delete a task (locked).
 * @param {string} taskId
 * @returns {Promise<{ success: boolean, error?: string }>}
 */
export async function remove(taskId) {
  return mutex.runExclusive(async () => {
    const tasks = await TaskRepo.loadTasks();
    const filtered = tasks.filter((t) => t.id !== taskId);
    if (filtered.length === tasks.length) return { success: false, error: 'Task not found' };
    await TaskRepo.saveTasks(filtered);
    return { success: true };
  });
}

export default {
  create,
  getById,
  list,
  update,
  remove,
};
